### PR TITLE
Started using StaticArrays, made a few breaking changes, towards automatic differentiation

### PR DESCRIPTION
**Pull request by zsunberg**
Hi guys,

I finished integrating Vec with StaticArrays, now every AbstractVec is also a StaticArray and hence an AbstractArray.

Along the way, I made some decisions about the interface and made some breaking changes to fix them. I could definitely use your feedback on whether they were good decisions. I think these should be pretty loud changes - i.e. they will cause errors rather than outputting incorrect results silently.

First, norm and abs (and a few others) were not used to mean the same thing as they do in base, so I made the following changes so that things are now consistent with the base library
1. norm -> normalize
2. abs -> norm
3. abs2 -> normsquared

I also got rid of the non-standard operator overloading for VecSE2 (the standard operators are automatically overloaded by StaticArrays) and added scale_euclidean() and normalize_euclidean() to be more verbose about operating only on the Euclidean part of the vector.

I also made some progress towards integration with ForwardDiff. It already works for a few things, but any real usefulness is being held up by JuliaDiff/ForwardDiff.jl#305 .

@MaximeBouton @kylejbrown17 This is a big PR, so we definitely don't need to pore over every line,  but I think our driving code will be higher quality if we review each other's stuff a little bit. @tawheeler your advice is always welcome as well.

### CODE REVIEW

File: src/common.jl [hunk 1]
```diff
@@ -0,0 +1,11 @@
+Base.isfinite(a::AbstractVec) = all(isfinite.(a))
+Base.isinf(a::AbstractVec) = any(isinf.(a))
+
+function StaticArrays.similar_type(::Type{V}, ::Type{Float64}, size::Size) where {V<:AbstractVec}
```
Comment (rejuvyesh) [inline src/common.jl]: Why not:
```julia
function StaticArrays.similar_type(::Type{V}, ::Type{F}, size::Size) where {V<:AbstractVec, F<:AbstractFloat}
    if size == Size(V) && eltype(V) == F
        return V
    end
end
```
That should atleast work with all kinds of floats?

**Review by MaximeBouton — COMMENTED** [review]
Awesome changes, we should probably still wait for ForwardDiff to work before merging. 
I am wondering if using `abs` will cause a method error or not.

File: src/common.jl (hunk 1 above)
Comment (zsunberg) [reply src/common.jl]: good call - implemented this. There were some weird things going on when I tried to do it for all `Real`. Will fix that eventually.

**Comment by zsunberg** [comment]
`abs` now throws an error.  and I implemented Jayesh's suggestion.

@MaximeBouton, I don't think there's any reason to wait for ForwardDiff because that didn't work before. The only reason to wait would be that it breaks people's code. I put a note in the README to show them how to get the old code.

I'll merge tomorrow unless someone has objections.

**Comment by MaximeBouton** [comment]
Ok that sounds good!

**Comment by rejuvyesh** [comment]
@zsunberg Remember to remove the `tags` file before you merge!

**Pull request merged**